drape_frontend/text_shape.cpp

```diff
@@ -236,47 +236,51 @@ void TextShape::Draw(ref_ptr<dp::GraphicsContext> context, ref_ptr<dp::Batcher>
-    primaryLayout = StraightTextLayout(strings::MakeUniString(titleDecl.m_primaryText), newFontSize,
-                                       titleDecl.m_primaryTextFont.m_isSdf, textures,
-                                       titleDecl.m_anchor, titleDecl.m_forceNoWrap);
+    if (auto const y = primaryLayout.GetPixelSize().y; y >= m_params.m_limits.y)
+    {
+      float const newFontSize = titleDecl.m_primaryTextFont.m_size * m_params.m_limits.y / y;
+      primaryLayout = StraightTextLayout(strings::MakeUniString(titleDecl.m_primaryText), newFontSize,
+                                         titleDecl.m_primaryTextFont.m_isSdf, textures,
+                                         titleDecl.m_anchor, titleDecl.m_forceNoWrap);
+    }
   }
 
-  drape_ptr<StraightTextLayout> secondaryLayout;
-  if (!titleDecl.m_secondaryText.empty())
+  glsl::vec2 primaryOffset;
+  glsl::vec2 secondaryOffset;
+  if (titleDecl.m_secondaryText.empty())
   {
-    secondaryLayout = make_unique_dp<StraightTextLayout>(
+    CalculateTextOffsets(titleDecl, primaryLayout.GetPixelSize(), m2::PointF{0.f, 0.f}, primaryOffset, secondaryOffset);
+    primaryOffset += glsl::vec2{m_symbolOffset.x, m_symbolOffset.y};
+  }
+  else
+  {
+    StraightTextLayout secondaryLayout{
         strings::MakeUniString(titleDecl.m_secondaryText), titleDecl.m_secondaryTextFont.m_size,
         titleDecl.m_secondaryTextFont.m_isSdf, textures, titleDecl.m_anchor,
-        titleDecl.m_forceNoWrap);
-  }
+        titleDecl.m_forceNoWrap};
 
-  glsl::vec2 primaryOffset;
-  glsl::vec2 secondaryOffset;
-  CalculateTextOffsets(titleDecl, primaryLayout.GetPixelSize(),
-                       secondaryLayout != nullptr ? secondaryLayout->GetPixelSize() : m2::PointF(0.0f, 0.0f),
-                       primaryOffset, secondaryOffset);
-  primaryOffset += glsl::vec2(m_symbolOffset.x, m_symbolOffset.y);
-  secondaryOffset += glsl::vec2(m_symbolOffset.x, m_symbolOffset.y);
+    if (secondaryLayout.GetGlyphCount() > 0)
+    {
+      CalculateTextOffsets(titleDecl, primaryLayout.GetPixelSize(),
+                         secondaryLayout.GetPixelSize(), primaryOffset, secondaryOffset);
+      secondaryOffset += glsl::vec2(m_symbolOffset.x, m_symbolOffset.y);
+      DrawSubString(context, secondaryLayout, titleDecl.m_secondaryTextFont, secondaryOffset, batcher,
+                    textures, false /* isPrimary */, titleDecl.m_secondaryOptional);
+    }
+  }
 
   if (primaryLayout.GetGlyphCount() > 0)
-  {
     DrawSubString(context, primaryLayout, titleDecl.m_primaryTextFont, primaryOffset, batcher,
                   textures, true /* isPrimary */, titleDecl.m_primaryOptional);
-  }
-
-  if (secondaryLayout != nullptr && secondaryLayout->GetGlyphCount() > 0)
-  {
-    DrawSubString(context, *secondaryLayout.get(), titleDecl.m_secondaryTextFont, secondaryOffset, batcher,
-                  textures, false /* isPrimary */, titleDecl.m_secondaryOptional);
-  }
 }
```

Comment:
Hm, its difficult to compare before/after now, but I suspect that ```DrawSubString``` order matters and its changed.

Reply:
1. I don't see any visual artifacts.
2. Texts should not overlap anyway.
3. Internally, geometry is batched.
4. Added a clarifying comment. We can roll it back if any artifacts arise. The overhead from unnecessary memory allocations is significant here.

Reply:
Honestly, I don't understand what performance became better here. The logic now is more complicated at least because of 2 similar function calls.

Reply:
Which 2 similar function calls?

New version does not have two memory allocations.